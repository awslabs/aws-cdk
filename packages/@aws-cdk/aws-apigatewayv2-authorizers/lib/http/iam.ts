import { HttpRouteAuthorizerConfig, IHttpRouteAuthorizer } from '@aws-cdk/aws-apigatewayv2';

/**
 * Authorize Http Api routes with IAM using sigv4 to sign request
 */
export class HttpIamAuthorizer implements IHttpRouteAuthorizer {
  public bind(): HttpRouteAuthorizerConfig {
    return {
      authorizationType: 'AWS_IAM',
    };
  }
}
