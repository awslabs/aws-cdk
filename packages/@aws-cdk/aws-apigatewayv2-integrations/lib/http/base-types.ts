import { HttpMethod, IVpcLink } from '@aws-cdk/aws-apigatewayv2';
import { Duration } from '@aws-cdk/core';

/**
 * Base options for private integration
 */
export interface HttpPrivateIntegrationOptions {
  /**
   * The vpc link to be used for the private integration
   *
   * @default - a new VpcLink is created
   */
  readonly vpcLink?: IVpcLink;

  /**
   * The HTTP method that must be used to invoke the underlying HTTP proxy.
   * @default HttpMethod.ANY
   */
  readonly method?: HttpMethod;
}

/**
 * Common properties to initialize a new `HttpProxyIntegration`.
 */
export interface CommonIntegrationProps {

  /**
   * The description of the integration
   *
   * @default - undefined
   */
  readonly description?: string;

  /**
   * Custom timeout for HTTP APIs
   *
   * @default - undefined
   */
  readonly timeout?: Duration;
}