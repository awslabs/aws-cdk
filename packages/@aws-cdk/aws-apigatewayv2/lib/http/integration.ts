import { Resource } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnIntegration } from '../apigatewayv2.generated';
import { IIntegration, AwsServiceIntegrationSubtype } from '../common';
import { IHttpApi } from './api';
import { HttpMethod, IHttpRoute } from './route';

// v2 - keep this import as a separate section to reduce merge conflict when forward merging with the v2 branch.
// eslint-disable-next-line
import { Construct as CoreConstruct } from '@aws-cdk/core';

/**
 * Represents an Integration for an HTTP API.
 */
export interface IHttpIntegration extends IIntegration {
  /** The HTTP API associated with this integration */
  readonly httpApi: IHttpApi;
}

/**
 * Supported integration types
 */
export enum HttpIntegrationType {
  /**
   * Integration type is a Lambda proxy
   * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html
   */
  LAMBDA_PROXY = 'AWS_PROXY',
  /**
   * Integration type is an HTTP proxy
   * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html
   */
  HTTP_PROXY = 'HTTP_PROXY',
  /**
   * Integration type is an AWS service
   * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-aws-services.html
   */
  AWS_PROXY = 'AWS_PROXY',
}

/**
 * Supported connection types
 */
export enum HttpConnectionType {
  /**
   * For private connections between API Gateway and resources in a VPC
   */
  VPC_LINK = 'VPC_LINK',
  /**
   * For connections through public routable internet
   */
  INTERNET = 'INTERNET',
}

/**
 * Payload format version for lambda proxy integration
 * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html
 */
export class PayloadFormatVersion {
  /** Version 1.0 */
  public static readonly VERSION_1_0 = new PayloadFormatVersion('1.0');
  /** Version 2.0 */
  public static readonly VERSION_2_0 = new PayloadFormatVersion('2.0');

  /**
   * A custom payload version.
   * Typically used if there is a version number that the CDK doesn't support yet
   */
  public static custom(version: string) {
    return new PayloadFormatVersion(version);
  }

  /** version as a string */
  public readonly version: string;

  private constructor(version: string) {
    this.version = version;
  }
}

/**
 * The integration properties
 */
export interface HttpIntegrationProps {
  /**
   * The HTTP API to which this integration should be bound.
   */
  readonly httpApi: IHttpApi;

  /**
   * Integration type
   */
  readonly integrationType: HttpIntegrationType;

  /**
   * Integration subtype
   */
  readonly integrationSubtype?: AwsServiceIntegrationSubtype;

  /**
   * Request parameters for integration
   */
  readonly requestParameters?: { [key: string]: any };

  /**
   * Integration URI.
   * This will be the function ARN in the case of `HttpIntegrationType.LAMBDA_PROXY`,
   * or HTTP URL in the case of `HttpIntegrationType.HTTP_PROXY`.
   */
  readonly integrationUri?: string;

  /**
   * The HTTP method to use when calling the underlying HTTP proxy
   * @default - none. required if the integration type is `HttpIntegrationType.HTTP_PROXY`.
   */
  readonly method?: HttpMethod;

  /**
   * The ID of the VPC link for a private integration. Supported only for HTTP APIs.
   *
   * @default - undefined
   */
  readonly connectionId?: string;

  /**
   * The type of the network connection to the integration endpoint
   *
   * @default HttpConnectionType.INTERNET
   */
  readonly connectionType?: HttpConnectionType;

  /**
   * `AWS::ApiGatewayV2::Api.CredentialsArn`
   *
   * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigatewayv2-api.html#cfn-apigatewayv2-api-credentialsarn
   */
  readonly credentialsArn?: string;

  /**
   * The version of the payload format
   * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html
   * @default - defaults to latest in the case of HttpIntegrationType.LAMBDA_PROXY`, irrelevant otherwise.
   */
  readonly payloadFormatVersion?: PayloadFormatVersion;
}

/**
 * The integration for an API route.
 * @resource AWS::ApiGatewayV2::Integration
 */
export class HttpIntegration extends Resource implements IHttpIntegration {
  public readonly integrationId: string;

  public readonly httpApi: IHttpApi;

  constructor(scope: Construct, id: string, props: HttpIntegrationProps) {
    super(scope, id);

    const { integrationType, integrationSubtype, requestParameters } = props;

    if (integrationType === HttpIntegrationType.AWS_PROXY && integrationSubtype && !requestParameters) {
      throw new Error('Must define request parameters for AWS service integration');
    }

    const integ = new CfnIntegration(this, 'Resource', {
      apiId: props.httpApi.httpApiId,
      connectionId: props.connectionId,
      connectionType: props.connectionType,
      credentialsArn: props.credentialsArn,
      integrationMethod: props.method,
      integrationType: props.integrationType,
      integrationSubtype: props.integrationSubtype,
      integrationUri: props.integrationUri,
      payloadFormatVersion: props.payloadFormatVersion?.version,
      requestParameters: this.renderRequestParameters(props.integrationSubtype, props.requestParameters),
    });
    this.integrationId = integ.ref;
    this.httpApi = props.httpApi;
  }

  private renderRequestParameters(
    integrationSubtype?: AwsServiceIntegrationSubtype,
    requestParameters?: { [key: string]: any },
  ): { [key: string]: any } | undefined {
    if (!integrationSubtype && !requestParameters) {
      return undefined;
    }
    if (!integrationSubtype) {
      throw new Error('Not rendering request parameters for non-existent AWS service integration');
    }
    if (integrationSubtype && !requestParameters) {
      throw new Error('Must define request parameters for AWS service integration');
    }
    switch (integrationSubtype) {
      case AwsServiceIntegrationSubtype.EVENT_BRIDGE_PUT_EVENTS:
        return {
          Detail: requestParameters?.detail,
          DetailType: requestParameters?.detailType,
          EventBusName: requestParameters?.eventBusName,
          Resources: requestParameters?.resources,
          Region: requestParameters?.region,
          Source: requestParameters?.source,
          Time: requestParameters?.time,
        };
      default:
        throw new Error(`Unknown integration subtype: ${integrationSubtype}`);
    }
  }
}

/**
 * Options to the HttpRouteIntegration during its bind operation.
 */
export interface HttpRouteIntegrationBindOptions {
  /**
   * The route to which this is being bound.
   */
  readonly route: IHttpRoute;

  /**
   * The current scope in which the bind is occurring.
   * If the `HttpRouteIntegration` being bound creates additional constructs,
   * this will be used as their parent scope.
   */
  readonly scope: CoreConstruct;
}

/**
 * The interface that various route integration classes will inherit.
 */
export interface IHttpRouteIntegration {
  /**
   * Bind this integration to the route.
   */
  bind(options: HttpRouteIntegrationBindOptions): HttpRouteIntegrationConfig;
}

/**
 * Config returned back as a result of the bind.
 */
export interface HttpRouteIntegrationConfig {
  /**
   * Integration type.
   */
  readonly type: HttpIntegrationType;

  /**
   * Integration URI
   */
  readonly uri?: string;

  /**
   * The HTTP method that must be used to invoke the underlying proxy.
   * Required for `HttpIntegrationType.HTTP_PROXY`
   * @default - undefined
   */
  readonly method?: HttpMethod;

  /**
   * The ID of the VPC link for a private integration. Supported only for HTTP APIs.
   *
   * @default - undefined
   */
  readonly connectionId?: string;

  /**
   * The type of the network connection to the integration endpoint
   *
   * @default HttpConnectionType.INTERNET
   */
  readonly connectionType?: HttpConnectionType;

  /**
   * The type of the network connection to the integration endpoint
   *
   * @default HttpConnectionType.INTERNET
   */
  readonly credentialsArn?: string;

  /**
   * Payload format version in the case of lambda proxy integration
   * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html
   * @default - undefined
   */
  readonly payloadFormatVersion: PayloadFormatVersion;

  /**
   * Integration subtype
   */
  readonly integrationSubtype?: AwsServiceIntegrationSubtype;

  /**
   * Request parameters for integration
   */
  readonly requestParameters?: Record<string, any>;
}
