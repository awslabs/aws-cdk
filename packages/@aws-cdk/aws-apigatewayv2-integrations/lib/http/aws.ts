import { IRole } from '@aws-cdk/aws-iam';
import { IStateMachine } from '@aws-cdk/aws-stepfunctions';
import { AwsServiceIntegration, AwsServiceIntegrationProps } from './private/integration';

/**
 * The common Step Functions integration resource for HTTP API
 */
abstract class StepFunctionsIntegration extends AwsServiceIntegration {

  /**
   *
   * @internal
   */
  protected _integrationService(): string {
    return 'StepFunctions';
  }
}

/**
 * Step Functions StartExecution integration properties
 */
export interface StepFunctionsStartExecutionIntegrationProps extends AwsServiceIntegrationProps {

  /**
   * The state machine to be executed
   */
  readonly stateMachine: IStateMachine;

  /**
   * The execution name
   *
   * @default - undefined
   */
  readonly name?: string;

  /**
   * The input parameters of execution
   *
   * @default - undefined
   */
  readonly input?: any;

  /**
   * The region of state machine
   *
   * @default - undefined
   */
  readonly region?: string;
}

/**
 * The StepFunctions-StartExecution integration resource for HTTP API
 */
export class StepFunctionsStartExecutionIntegration extends StepFunctionsIntegration {

  constructor(private readonly _props: StepFunctionsStartExecutionIntegrationProps) {
    super(_props);
  }

  /**
   *
   * @internal
   */
  protected _integrationAction(): string {
    return 'StartExecution';
  }

  /**
   *
   * @internal
   */
  protected _fulfillRole(credentialsRole: IRole): void {
    this._props.stateMachine.grantStartExecution(credentialsRole);
  }

  /**
   *
   * @internal
   */
  protected _buildRequestParameters(): { [key: string]: any } {
    return {
      StateMachineArn: this._props.stateMachine.stateMachineArn,
      Name: this._props.name,
      Input: this._props.input,
      Region: this._props.region,
    };
  }
}

/**
 * Step Functions StartSyncExecution integration properties
 */
export interface StepFunctionsStartSyncExecutionIntegrationProps extends StepFunctionsStartExecutionIntegrationProps {

  /**
   * Passes the AWS X-Ray trace header. The trace header can also be passed in the request payload.
   *
   * @default - undefined
   */
  readonly traceHeader?: string;
}

/**
 * The StepFunctions-StartExecution integration resource for HTTP API
 */
export class StepFunctionsStartSyncExecutionIntegration extends StepFunctionsIntegration {

  constructor(private readonly _props: StepFunctionsStartSyncExecutionIntegrationProps) {
    super(_props);
  }

  /**
   *
   * @internal
   */
  protected _integrationAction(): string {
    return 'StartSyncExecution';
  }

  /**
   *
   * @internal
   */
  protected _fulfillRole(credentialsRole: IRole): void {
    this._props.stateMachine.grantStartExecution(credentialsRole);
  }

  /**
   *
   * @internal
   */
  protected _buildRequestParameters(): { [key: string]: any } {
    return {
      StateMachineArn: this._props.stateMachine.stateMachineArn,
      Name: this._props.name,
      Input: this._props.input,
      Region: this._props.region,
      TraceHeader: this._props.traceHeader,
    };
  }
}