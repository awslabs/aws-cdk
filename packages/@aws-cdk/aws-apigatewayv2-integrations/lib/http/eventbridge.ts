import {
  HttpRouteIntegrationBindOptions,
  HttpRouteIntegrationConfig,
  AwsServiceIntegrationSubtype,
} from '@aws-cdk/aws-apigatewayv2';
import { IEventBus } from '@aws-cdk/aws-events';
import { PolicyStatement, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { AwsServiceIntegration } from './private/integration';


export enum EventBridgeTargetAction {
  PUT_EVENTS = 'EventBridge-PutEvents'
}

/**
 * Properties to initialize `EventBridgeIntegration`.
 */
export interface EventBridgeIntegrationProps {
  /**
  * The EventBridge API action to integrate.
  */
  readonly targetAction: EventBridgeTargetAction;

  /**
   * The target event bus.
   */
  readonly eventBus: IEventBus;

  /**
   * The event source.
   */
  readonly eventSource?: string;

  /**
   * The EventBridge PutEvents request parameters.
   * @see https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_PutEvents.html
   */
  readonly requestParameters: EventBridgeIntegrationRequestParameters;
}

/**
 * Integration request parameters for EventBridge
 * @see https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_PutEventsRequestEntry.html#eventbridge-Type-PutEventsRequestEntry-Detail
 */
export interface EventBridgeIntegrationRequestParameters {
  /**
  * A valid JSON string. There is no other schema imposed. The JSON string may contain fields and nested subobjects.
  *
  * @default none
  */
  readonly detail?: string;

  /**
  * Free-form string used to decide what fields to expect in the event detail.
  *
  * @default none
  */
  readonly detailType?: string;

  /**
  * The AWS region. NB: Particular to EventBridge-PutEvents API Gateway integration and not EventBridge's PutEvents API.
  *
  * @default none
  */
  readonly region?: string;

  /**
  * AWS resources, identified by Amazon Resource Name (ARN), which the event primarily concerns. Any number,
  * including zero, may be present.
  *
  * @default none
  */
  readonly resources?: string[];

  /**
  * The source of the event.
  *
  * @default none
  */
  readonly source?: string;

  /**
  * The time stamp of the event, per RFC3339. If no time stamp is provided, the time stamp of the PutEvents call is
  * used.
  *
  * @default none
  */
  readonly time?: string;
}

/**
 * The EventBridge integration resource for HTTP API
 */
export class EventBridgeIntegration extends AwsServiceIntegration {
  constructor(private readonly props: EventBridgeIntegrationProps) {
    super();
  }

  public bind(options: HttpRouteIntegrationBindOptions): HttpRouteIntegrationConfig {
    switch (this.props.targetAction) {
      case EventBridgeTargetAction.PUT_EVENTS:
        this.integrationSubtype = AwsServiceIntegrationSubtype.EVENT_BRIDGE_PUT_EVENTS;
        break;
      default:
        this.integrationSubtype = AwsServiceIntegrationSubtype.EVENT_BRIDGE_PUT_EVENTS;
    }

    const { scope } = options;

    const role = new Role(scope, 'EventBridgeIntegrationRole', {
      description: 'Role for API Gateway to publish Events',
      assumedBy: new ServicePrincipal('apigateway.amazonaws.com'),
    });

    role.addToPolicy(new PolicyStatement({
      actions: ['events:PutEvents'],
      resources: [this.props.eventBus.eventBusArn],
      conditions: (() => !('eventSource' in this.props) ? undefined : {
        StringEquals: {
          'events:source': this.props.eventSource,
        },
      })(),
    }));

    return {
      payloadFormatVersion: this.payloadFormatVersion,
      type: this.integrationType,
      connectionType: this.connectionType,
      credentialsArn: role.roleArn,
      integrationSubtype: this.integrationSubtype,
      requestParameters: {
        ...this.props.requestParameters,
        eventBusName: this.props.eventBus.eventBusName,
      },
    };
  }
}
