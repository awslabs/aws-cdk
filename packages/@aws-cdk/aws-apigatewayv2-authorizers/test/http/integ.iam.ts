import * as path from 'path';
import { HttpApi, HttpMethod } from '@aws-cdk/aws-apigatewayv2';
import { LambdaProxyIntegration } from '@aws-cdk/aws-apigatewayv2-integrations';
import { User, CfnAccessKey, ManagedPolicy } from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import { App, Stack, CfnOutput } from '@aws-cdk/core';
import { HttpIamAuthorizer } from '../../lib/http/iam';

/*
 * Stack verification steps:
 * * `curl <url>` should return 403
 * * `curl -H 'Authorization: ' -H 'x-amz-date: ' -H 'Accept: application/json' <url>` should return 200
 * @see [../integ.iam.signature/README.md] to generate header values using this stack's outputs
*/

const app = new App();
const stack = new Stack(app, 'IAMAuthorizerInteg');

const httpApi = new HttpApi(stack, 'MyHttpApi');

const authorizer = new HttpIamAuthorizer();

const handler = new lambda.Function(stack, 'lambda', {
  runtime: lambda.Runtime.NODEJS_12_X,
  handler: 'index.handler',
  code: lambda.AssetCode.fromAsset(path.join(__dirname, '../integ.lambda.handler')),
});

httpApi.addRoutes({
  path: '/',
  methods: [HttpMethod.GET],
  integration: new LambdaProxyIntegration({ handler }),
  authorizer,
});

const user = new User(stack, 'test-user');

user.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('AmazonAPIGatewayInvokeFullAccess'));

const accessKey = new CfnAccessKey(stack, 'access-key', {
  userName: user.userName,
});

new CfnOutput(stack, 'api_url', { value: httpApi.url! });
new CfnOutput(stack, 'access_key', { value: accessKey.ref });
new CfnOutput(stack, 'secret_access_key', { value: accessKey.attrSecretAccessKey });