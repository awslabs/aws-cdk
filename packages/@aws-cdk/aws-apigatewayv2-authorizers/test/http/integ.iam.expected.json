{
  "Resources": {
    "MyHttpApi8AEAAC21": {
      "Type": "AWS::ApiGatewayV2::Api",
      "Properties": {
        "Name": "MyHttpApi",
        "ProtocolType": "HTTP"
      }
    },
    "MyHttpApiDefaultStageDCB9BC49": {
      "Type": "AWS::ApiGatewayV2::Stage",
      "Properties": {
        "ApiId": {
          "Ref": "MyHttpApi8AEAAC21"
        },
        "StageName": "$default",
        "AutoDeploy": true
      }
    },
    "MyHttpApiGETIAMAuthorizerIntegMyHttpApiGET271B2CE5PermissionE3A1E0E1": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "lambda8B5974B5",
            "Arn"
          ]
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition"
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region"
              },
              ":",
              {
                "Ref": "AWS::AccountId"
              },
              ":",
              {
                "Ref": "MyHttpApi8AEAAC21"
              },
              "/*/*/"
            ]
          ]
        }
      }
    },
    "MyHttpApiGETHttpIntegration6f095b8469365f72e33fa33d9711b140516EBE31": {
      "Type": "AWS::ApiGatewayV2::Integration",
      "Properties": {
        "ApiId": {
          "Ref": "MyHttpApi8AEAAC21"
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::GetAtt": [
            "lambda8B5974B5",
            "Arn"
          ]
        },
        "PayloadFormatVersion": "2.0"
      }
    },
    "MyHttpApiGETE0EFC6F8": {
      "Type": "AWS::ApiGatewayV2::Route",
      "Properties": {
        "ApiId": {
          "Ref": "MyHttpApi8AEAAC21"
        },
        "RouteKey": "GET /",
        "AuthorizationType": "AWS_IAM",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "MyHttpApiGETHttpIntegration6f095b8469365f72e33fa33d9711b140516EBE31"
              }
            ]
          ]
        }
      }
    },
    "lambdaServiceRole494E4CA6": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "lambda8B5974B5": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters1fd1c15cb7d5e2e36a11745fd10b4b7c3ca8eb30642b41954630413d2b913cdaS3Bucket2E6D85D3"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters1fd1c15cb7d5e2e36a11745fd10b4b7c3ca8eb30642b41954630413d2b913cdaS3VersionKey22B8E7C6"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters1fd1c15cb7d5e2e36a11745fd10b4b7c3ca8eb30642b41954630413d2b913cdaS3VersionKey22B8E7C6"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Role": {
          "Fn::GetAtt": [
            "lambdaServiceRole494E4CA6",
            "Arn"
          ]
        },
        "Handler": "index.handler",
        "Runtime": "nodejs12.x"
      },
      "DependsOn": [
        "lambdaServiceRole494E4CA6"
      ]
    },
    "testuser14267055": {
      "Type": "AWS::IAM::User",
      "Properties": {
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/AmazonAPIGatewayInvokeFullAccess"
              ]
            ]
          }
        ]
      }
    },
    "accesskey": {
      "Type": "AWS::IAM::AccessKey",
      "Properties": {
        "UserName": {
          "Ref": "testuser14267055"
        }
      }
    }
  },
  "Parameters": {
    "AssetParameters1fd1c15cb7d5e2e36a11745fd10b4b7c3ca8eb30642b41954630413d2b913cdaS3Bucket2E6D85D3": {
      "Type": "String",
      "Description": "S3 bucket for asset \"1fd1c15cb7d5e2e36a11745fd10b4b7c3ca8eb30642b41954630413d2b913cda\""
    },
    "AssetParameters1fd1c15cb7d5e2e36a11745fd10b4b7c3ca8eb30642b41954630413d2b913cdaS3VersionKey22B8E7C6": {
      "Type": "String",
      "Description": "S3 key for asset version \"1fd1c15cb7d5e2e36a11745fd10b4b7c3ca8eb30642b41954630413d2b913cda\""
    },
    "AssetParameters1fd1c15cb7d5e2e36a11745fd10b4b7c3ca8eb30642b41954630413d2b913cdaArtifactHash82A279EA": {
      "Type": "String",
      "Description": "Artifact hash for asset \"1fd1c15cb7d5e2e36a11745fd10b4b7c3ca8eb30642b41954630413d2b913cda\""
    }
  },
  "Outputs": {
    "apiurl": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "MyHttpApi8AEAAC21"
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region"
            },
            ".",
            {
              "Ref": "AWS::URLSuffix"
            },
            "/"
          ]
        ]
      }
    },
    "accesskey": {
      "Value": {
        "Ref": "accesskey"
      }
    },
    "secretaccesskey": {
      "Value": {
        "Fn::GetAtt": [
          "accesskey",
          "SecretAccessKey"
        ]
      }
    }
  }
}